File: backend/dtos/src/main/java/io/mateu/dtos/Crud.java
package io.mateu.dtos;

import java.util.Collections;
import java.util.List;

/**
 * A crud
 *
 * @param listId An id for this crud
 * @param title The title
 * @param subtitle The subtitle
 * @param canEdit If records can be edited. This make the Edit button visible
 * @param searchForm The form for searching. It can include several filters
 * @param columns The columns
 * @param actions The actions which can be run for this crud. They end up as buttons
 */
public record Crud(
    String listId,
    String title,
    String subtitle,
    boolean canEdit,
    SearchForm searchForm,
    List<Column> columns,
    List<Action> actions)
    implements ComponentMetadata {

  public Crud {
    columns = Collections.unmodifiableList(columns);
    actions = Collections.unmodifiableList(actions);
  }

  @Override
  public List<Column> columns() {
    return Collections.unmodifiableList(columns);
  }

  @Override
  public List<Action> actions() {
    return Collections.unmodifiableList(actions);
  }
}

File: frontend/web/vaadin/src/mateu/shared/apiClients/dtos/componentmetadata/Crud.ts
import ComponentMetadata from "@mateu/shared/apiClients/dtos/ComponentMetadata";
import FormField from "@mateu/shared/apiClients/dtos/componentmetadata/FormField";
import Action from "@mateu/shared/apiClients/dtos/componentmetadata/Action";
import Table from "@mateu/shared/apiClients/dtos/componentmetadata/Table";

export default interface Crud extends ComponentMetadata {

    id: string
    title: string
    subtitle: string
    searchable: boolean
    filters: FormField[]
    table: Table
    actions: Action[]
    pageSize: number,
    clientSidePagination: boolean
    searchOnEnter: boolean
    autoFocusOnSearchText: boolean

}